tests/End2EndTestCase.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/**
+ * This file is part of the contentful.php package.
+ *
+ * @copyright 2015-2017 Contentful GmbH
+ * @license   MIT
+ */
+
+namespace Contentful\Tests\Delivery;
+
+use Contentful\Delivery\Client;
+use Contentful\Log\ArrayLogger;
+use Symfony\Component\Filesystem\Filesystem;
+
+class End2EndTestCase extends \PHPUnit_Framework_TestCase
+{
+    const CACHE_DIR = __DIR__ . '/../build/cache';
+
+    /**
+     * @var Client[]
+     */
+    private static $clients = [];
+
+    public static function setUpBeforeClass()
+    {
+        self::$clients = [
```

Comment:
The CMA client only needs a token, so that's why that approach works there. Here we need a bunch of different clients with very different requirements, so whereas the CMA gets away with `getReadOnlySpaceManager`/`getReadWriteSpaceManager`, here we'd need 9 different getter methods, or one with a huge switch/case, where the client would be instantiated on-demand and then saved.

But as the test suite is most often execute as a whole, it actually make sense to have the clients defined like this (instead of created/destroyed multiple times), and I could actually improve on this by checking that if the array is not empty, then returning early because clients are already set up.